pkg/auth/auth_test.go

```diff
@@ -0,0 +1,390 @@
+package auth_test
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/fastly/cli/pkg/app"
+	"github.com/fastly/cli/pkg/auth"
+	"github.com/fastly/cli/pkg/config"
+	"github.com/fastly/cli/pkg/manifest"
+	"github.com/fastly/cli/pkg/mock"
+	"github.com/fastly/cli/pkg/testutil"
+	"github.com/fastly/go-fastly/v6/fastly"
+)
+
+// TestAuthSuccess validates we're able to initialise the authentication flow
+// and to get a new access token returned and persisted to disk.
+func TestAuthSuccess(t *testing.T) {
+	wd, root := createTestEnvironment(t)
+	defer os.RemoveAll(root)
+	defer os.Chdir(wd)
+
+	var stdout bytes.Buffer
+	args := testutil.Args("pops")
+	opts := testutil.NewRunOpts(args, &stdout)
+	opts.ConfigPath = filepath.Join(root, manifest.Filename)
+	opts.ClientFactory = mock.ClientFactory(mockAPISuccess)
+	opts.Stdin = strings.NewReader("y") // Authorise opening of web browser.
+
+	endpoint := make(chan string)
+	generatedToken := "123"
+	auth.Browser = mockBrowser(endpoint, generatedToken)
+	opts.AuthService = <-endpoint
+
+	err := app.Run(opts)
+	if err != nil {
+		t.Log(stdout.String())
+		t.Fatal(err)
+	}
+
+	wantOutput := "We are about to initialise a new authentication flow"
+	testutil.AssertStringContains(t, stdout.String(), wantOutput)
+
+	var f config.File
+	err = f.Read(opts.ConfigPath, opts.Stdin, opts.Stdout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wantToken := "123"
+	if f.User.Token != wantToken {
+		t.Errorf("want token '%s', have '%s'", wantToken, f.User.Token)
+	}
+	wantEmail := "test@example.com"
+	if f.User.Email != wantEmail {
+		t.Errorf("want email '%s', have '%s'", wantEmail, f.User.Email)
+	}
+}
+
+// TestAuthTokenFlag validates the use of --token will result in no
+// authentication flow being initialised, as well as the token not being
+// persisted to disk.
+func TestAuthTokenFlag(t *testing.T) {
+	wd, root := createTestEnvironment(t)
+	defer os.RemoveAll(root)
+	defer os.Chdir(wd)
+
+	var stdout bytes.Buffer
+	args := testutil.Args("pops --token 123")
+	opts := testutil.NewRunOpts(args, &stdout)
+	opts.ConfigPath = filepath.Join(root, manifest.Filename)
+	opts.ClientFactory = mock.ClientFactory(mockAPISuccess)
+
+	endpoint := make(chan string)
+	generatedToken := "123"
+	auth.Browser = mockBrowser(endpoint, generatedToken)
+	opts.AuthService = <-endpoint
+
+	err := app.Run(opts)
+	if err != nil {
+		t.Log(stdout.String())
+		t.Fatal(err)
+	}
+
+	wantToIgnore := "We are about to initialise a new authentication flow"
+	testutil.AssertStringDoesntContain(t, stdout.String(), wantToIgnore)
+
+	var f config.File
+	err = f.Read(opts.ConfigPath, opts.Stdin, opts.Stdout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if f.User.Token != "" {
+		t.Errorf("want no token, have '%s'", f.User.Token)
+	}
+	if f.User.Email != "" {
+		t.Errorf("want no email, have '%s'", f.User.Email)
+	}
+}
```

Comment:
Here is where we validate the use of `--token` skips the authentication flow but also that the given token doesn't accidentally get persisted to disk (we only persist the token to disk if we've either gone through the authentication flow or if the user has run the `fastly configure` command).